### ISSUE TITLE
Markers lazy loading

### ISSUE THREAD

**Issue by Nohus** [p-lr/MapCompose#46]
I am seeing performance issues when using markers:
```
I/Choreographer: Skipped 409 frames!  The application may be doing too much work on its main thread.
```
This is when adding markers.
Spent 1.64s inside `MarkerComposer.kt`, out of which 333ms in my marker composable.

I have 200 markers on the map, but I am doing marker [generalization](https://en.wikipedia.org/wiki/Cartographic_generalization), so more markers appear when you zoom in, but because you are viewing a smaller area the total amount of markers on screen rarely exceeds 5. However it seems the library is always rendering all of them? The problem is much worse the more I zoom in.

**Comment by p-lr** [comment]
I can confirm all markers are laid out (not rendered, this is different).
Based on it's visibility, a marker can be added/removed dynamically. Perhaps this could be managed by the library. This can also be  managed by the application, using the bounding box.
I would see such an improvement land as part of a 2.1.x version.

**Comment by p-lr** [comment]
The good thing is, it should be easy to reproduce in the demo and evaluate possible solutions.

**Comment by Nohus** [comment]
I have implemented this in application code based on `visibleArea()` and there are no performance issues anymore. :tada:

**Comment by TheBrisingers** [comment]
Hello,
First of all, congratulations for the work done and its quality.

I work on an application where I have to indicate several hundred markers on a map. My goal is to display/create only the markers located on the visible part of the screen (to improve performance). I saw that it was possible to retrieve the `visibleArea()` but I'm not sure how to use it to sort the markers to display.

Also, I haven't figured out how to get the current zoom level in order to set a boundary from which they will display.

I hope you can help me.

**Comment by p-lr** [comment]
The `visibleArea()` gives you the normalized coordinates of the 4 points which defines the visible area (they define a rectangle). To know whether a marker is visible, you just have to check if its normalized coordinates are within the rectangle. The mathematical formula isn't trivial but it could be added as a helper function.
Nevertheless, if you don't use the rotation, you don't need to use `visibleArea()` - you can use `visibleBoundingBox()` instead. This later api also gives you the visible area, but only if the map isn't rotated. In this case, it's trivial to check if a point belongs to the rectangle.
If you use the rotation feature, let me know.

**Comment by TheBrisingers** [comment]
Indeed I do not use the rotation, I will try with `visibleBoundingBox()` and I should succeed in filtering the display.

Regarding my second question, is it possible to know the current level of the displayed maps? Or is only the `scale` value available?

**Comment by p-lr** [comment]
Indeed, only the scale is available.

**Comment by timhughes89** [comment]
Hi, 

Thanks for the library so far especially for compose, is there any chance you can please provide an example of what you deem to be best practice for rending 100's of markers on screen?

**Comment by p-lr** [comment]
When I have time, I will add a demo with hundreds of markers and lazy loading technique.

**Comment by p-lr** [comment]
I'm working on it. However, I stumbled upon a glaring issue with markers. Even with a few dozens, there were noticeable stutters.
Finally, I found the culprit and just pushed the changes -- will release 2.0.5 soon.
On an average device, we can now render ~200 markers with acceptable performance. This is way more than needed to implement an efficient clustering algorithm (which I'm working on atm).
Even when performing markers generalization (clustering), 2.0.5 should show a noticeable performance improvement.

**Comment by p-lr** [comment]
I've added a new demo called "Markers clustering & lazy loading".
It uses an experimental density based clustering algorithm. Eventually, it may be part of the library. However, I'd like to have feedback or thoughts about the expected api.

**Comment by TheBrisingers** [comment]
Hello !

Your work is always of such high quality and I thank you for this demo.
I have noticed a simple little detail that seems easy to fix. Sometimes, because of the randomness, points end up really close together and despite zooming in to the maximum it doesn't manage to separate. I know there is a parameter to adjust the epsilon between the markers but a condition on the maximum zoom might be a good idea to add.
Besides that, I'm particularly interested in the "contains" function on the "visibleArea" and "referentialSnapshotFlow" of MapState. When do you plan to add these features to the library?

Thanks

**Comment by p-lr** [comment]
> I have noticed a simple little detail that seems easy to fix. Sometimes, because of the randomness, points end up really close together and despite zooming in to the maximum it doesn't manage to separate. I know there is a parameter to adjust the epsilon between the markers but a condition on the maximum zoom might be a good idea to add.

Yes. A scale threshold, above which the clustering would be disabled. I keep that in mind.

> When do you plan to add these features to the library?

Integrating this into the library requires a bit more work. One can imagine that not all markers should be eligible for clusterization. Some kind of markers could be clusterized using some style of cluster, while others would correspond to a different style of cluster.
Given what's already done, we're almost there. I can start working on it in a specific branch.

@Nohus Could you please have a look at the demo and tell me what you think? Do you believe that this could serve your use case?

**Comment by p-lr** [comment]
@Nohus Could you have a look at the demo and tell me what you think? Do you believe that this could serve your use case?

**Comment by p-lr** [comment]
See also #3

**Comment by p-lr** [comment]
@TimPushkin It's now in the library. See #3 for more details.
I haven't published new artifacts to mavencentral yet. The next version will be 2.1.0

**Comment by TimPushkin** [comment]
@p-lr Thanks, I'm actually more interested in lazy loading and now I got a little bit confused: so, will the next version have lazy loading only when clustering is enabled?

**Comment by TimPushkin** [comment]
@p-lr Yes, thank you for the clarifications! Another use case for lazy loading only is to just remove all (of most of) the markers when the map is zoomed out -- that's what I'm planning to do. It may sound quite similar to clustering, but no new markers representing clusters are added.

**Comment by p-lr** [comment]
Since a marker cannot be "lazy-loaded only" and managed by clusterer, I've created the concept of `rendering strategy`. Now, there are 3 strategies:
- Default (eagerly laid-out)
- Lazy-loading
- Clustering

The `rendering strategy` is set when adding a marker. For example:
```kotlin
state.addMarker(
    id = "marker-$i",
    x = 0.2,
    y = 0.3,
    renderingStrategy = RenderingStrategy.LazyLoading("default"),
) {
    Marker()
}
```

The api which was recently introduced for the clustering had to be touched a bit (after-all, it was declared experimental). Anyway, there's a new demo specifically for lazy-loading only, and the demo for clustering is updated.
As it turns out, the next 2.1.0 will have both clustering and lazy-loading only.

@TimPushkin  When removing a lazy-loader using `MapState.removeLazyLoader()` api, all markers managed by the lazy-loader are removed by default.

**Comment by Nohus** [comment]
@p-lr I am back now, sorry for the delay. I see I was mentioned but the comment was edited. Is there anything I can help with/test now?

**Comment by p-lr** [comment]
@Nohus After the initial clustering demo, I wasn't sure about which api we should have for this, so I asked your input. Eventually, I gave it a shot and proposed an experimental api a few days ago. But @TimPushkin made a good point: we should also be able to have lazy loading only. That's what I added today.
If you get the chance to test the new clustering and/or lazy-loading api, and give your feedback (perhaps some more options to have, or new apis, or issues you encountered), that would help a lot. Thanks!

**Comment by Nohus** [comment]
Sounds good, will test it all and report back tomorrow.

**Comment by p-lr** [comment]
I suggest we continue the discussion [there](https://github.com/p-lr/MapCompose/discussions/66), unless we want to add something relevant to this issue.

**Comment by p-lr** [comment]
Recent changes on master have improved the performance of markers even more. The clickable property of a marker no longer has a negative impact on performance.
I'm now personally satisfied with the end result, and if no feedback or bug report is made within the next week, I'll publish the 2.1.0 version.

**Comment by p-lr** [comment]
Implemented in 2.1.0
